File: app/src/main/java/com/codeprogression/mvpb/MainActivity.java
package com.codeprogression.mvpb;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import com.codeprogression.mvpb.core.PerActivity;

public class MainActivity extends AppCompatActivity {


    @PerActivity
    @dagger.Component()
    public interface Component{
        void inject(MainActivity view);
        void inject(MainView view);
    }

    public static Component component(){
        return component;
    }

    private static Component component;

    private void buildComponent() {
        if (component == null) {
            component = DaggerMainActivity_Component.builder()
                    .build();
            component.inject(this);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        buildComponent();
        setContentView(R.layout.main_view);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        component = null;
    }
}

File: app/src/main/java/com/codeprogression/mvpb/MainActivity.kt
package com.codeprogression.mvpb

import android.os.Bundle
import android.support.v7.app.AppCompatActivity

class MainActivity : AppCompatActivity() {


    private fun buildComponent() {
        if (component == null) {
            component = DaggerMainComponent.builder().build()
            component!!.inject(this)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        buildComponent()
        setContentView(R.layout.main_view)
    }

    override fun onDestroy() {
        super.onDestroy()
        component = null
    }

    companion object {

        fun component(): MainComponent? {
            return component
        }

        private var component: MainComponent? = null
    }
}
